src/formatting/imports.rs

```diff
@@ -168,7 +168,19 @@ pub(crate) fn merge_use_trees(use_trees: Vec<UseTree>, merge_by: SharedPrefix) -
             continue;
         }
 
-        for flattened in use_tree.flatten() {
+        for mut flattened in use_tree.flatten() {
+            if merge_by == SharedPrefix::Module {
```

Comment:
Any ideas why this isn't needed for the `Crate` variant?

Reply:
As I understand, the following code transforms ahead `foo::self` to `foo`: https://github.com/rust-lang/rustfmt/blob/0e908555273391bc3aba087abae72791451fd84a/src/formatting/imports.rs#L474-L477
Therefore, the only reason `foo::self` can be present in the code you asked about, is that `flatten()` created `foo::self` from `foo::{self, bar}`, in addition to creating `foo::bar`.  For `SharedPrefix::Module` these too imports will remain separate, so `{}` should be added to `self`.  However for `SharedPrefix::Crate` they will be merged again.  This is because how `share_prefix()` identify shared prefix: https://github.com/rust-lang/rustfmt/blob/0e908555273391bc3aba087abae72791451fd84a/src/formatting/imports.rs#L570-L573


When checking this I found two examples of imports formatting which I am not sure are correct.  The first is when using `imports_granularity = "Crate"`:
```rust
use foo;
use foo::{self};
```
is formatted as:
```rust
use foo::{
    self, {self},
};
```
Is that o.k. or it should have been `use foo::{self}`?

The second possible formatting issue is when using `imports_granularity = "Module"`:
```rust
use foo;
use bar;
```
is formatted as:
```rust
use {bar, foo};
```
Is this o.k.?